File: Lab1/app/src/main/java/ru/sfedu/sergeysh/lab1/ui/login/LoginViewModel.java
package ru.sfedu.sergeysh.lab1.ui.login;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import ru.sfedu.sergeysh.lab1.data.LoginRepository;
import ru.sfedu.sergeysh.lab1.R;

public class LoginViewModel extends ViewModel {

    private MutableLiveData<LoginFormState> loginFormState = new MutableLiveData<>();
    private MutableLiveData<Boolean> isLoggedIn = new MutableLiveData<>();
    private MutableLiveData<LoggedInUser> loggedInUser = new MutableLiveData<>();
    private LoginRepository loginRepository;

    LoginViewModel(LoginRepository loginRepository) {
        this.loginRepository = loginRepository;
    }

    LiveData<LoginFormState> getLoginFormState() {
        return loginFormState;
    }

    LiveData<Boolean> getIsLoggedIn() {
        return isLoggedIn;
    }

    LiveData<LoggedInUser> getLoggedInUser() {
        return loggedInUser;
    }

    public void login(String username, String password) {
        // can be launched in a separate asynchronous job
        LoggedInUser user = loginRepository.login(username, password);
        this.loggedInUser.setValue(user);
        this.isLoggedIn.setValue(user != null);
    }

    public void loginDataChanged(String username, String password) {
        Integer usernameError = isUsernameValid(username) ? null : R.string.invalid_username;
        Integer passwordError = isPasswordValid(password) ? null : R.string.invalid_password;
        if (usernameError != null || passwordError != null) {
            loginFormState.setValue(new LoginFormState(usernameError, passwordError));
        } else {
            loginFormState.setValue(new LoginFormState(true));
        }
    }

    // A placeholder username validation check
    private boolean isUsernameValid(String username) {
        return username != null && !username.trim().isEmpty();
    }

    // A placeholder password validation check
    private boolean isPasswordValid(String password) {
        return password != null && password.trim().length() > 5;
    }
}

File: Lab2/app/src/main/java/ru/sfedu/sergeysh/lab2/ui/login/LoginViewModel.kt
package ru.sfedu.sergeysh.lab2.ui.login

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.launch
import ru.sfedu.sergeysh.lab2.R
import ru.sfedu.sergeysh.lab2.data.login.LoggedInUser
import ru.sfedu.sergeysh.lab2.data.login.LoginRepository

class LoginViewModel(private val loginRepository: LoginRepository) : ViewModel() {

    private val _loginForm: MutableLiveData<LoginFormState> = MutableLiveData<LoginFormState>()
    val loginFormState: LiveData<LoginFormState> = _loginForm

    private val _loginResult: MutableLiveData<LoginResult> = MutableLiveData<LoginResult>()
    val loginResult: LiveData<LoginResult> = _loginResult

    fun login(username: String, password: String) {
        viewModelScope.launch {
            val user: LoggedInUser? = loginRepository.login(username, password)

            if (user != null) {
                _loginResult.value = LoginResult(success = user.displayName)
            } else {
                _loginResult.value = LoginResult(error = R.string.login_failed)
            }
        }
    }

    fun loginDataChanged(username: String, password: String) {
        if (!isUserNameValid(username)) {
            _loginForm.value = LoginFormState(usernameError = R.string.invalid_username)
        } else if (!isPasswordValid(password)) {
            _loginForm.value = LoginFormState(passwordError = R.string.invalid_password)
        } else {
            _loginForm.value = LoginFormState(isDataValid = true)
        }
    }

    // A placeholder username validation check
    private fun isUserNameValid(username: String): Boolean {
        return username.isNotBlank()
    }

    // A placeholder password validation check
    private fun isPasswordValid(password: String): Boolean {
        return password.length > 5
    }
}
